examples/memory_with_compression/_memory_with_compress.py

```diff
@@ -0,0 +1,587 @@
+                of messages to compress. If None (default), the
+                self.compress_func will be used.
+            compression_trigger_func (Callable[[List[Msg]], Awaitable[bool]]):
+                Optional function to trigger compression. If None (default),
+                the self.compression_trigger_func will be used.
+            memory (List[Msg] | None):
+                The memory to check and compress. If None (default), the
+                memory_storage will be used.
+
+        Returns:
+            tuple[bool, List[Msg]]: A tuple containing a boolean value
+                    indicating if compression was triggered and the
+                    compressed memory. The boolean value is True if
+                    compression was triggered, False otherwise. The
+                    compressed memory is the list of messages that were
+                    compressed. If compression was not triggered, the
+                    compressed memory is the same as the input memory.
+        """
+        # if memory is not provided, use the _memory
+        if memory is None:
+            memory = await self.memory_storage.get()
+        # if compress_func is not provided, use the self.compress_func
+        if compress_func is None:
+            compress_func = self.compress_func
+        # if compression_trigger_func is not provided, use the self.
+        # compression_trigger_func
+        if compression_trigger_func is None:
+            compression_trigger_func = self.compression_trigger_func
+
+        # check if the memory needs compression by compression_trigger_func
+        # and compress it if needed.
+        # Notice that compression_trigger_func is optional,
+        # so if it is not provided, the memory will not be compressed
+        # by compression_trigger_func.
+        if compression_trigger_func is not None:
+            should_compress = await compression_trigger_func(memory)
+        else:
+            should_compress = False
+
+        if should_compress:
+            compressed_memory = await compress_func(memory)
+        else:
+            compressed_memory = memory
+        return should_compress, compressed_memory
+
+    async def retrieve(self, *args: Any, **kwargs: Any) -> None:
+        """
+        Retrieve items from the memory.
+        This method is not implemented as get_memory is used instead.
+        """
+        raise NotImplementedError(
+            "Use get_memory() instead of retrieve() for MemoryWithCompress",
+        )
+
+    async def size(self) -> int:
+        """
+        Get the size of the memory.
+
+        Returns:
+            int: The number of messages in chat_history_storage.
+        """
+        return len(await self.chat_history_storage.get())
+
+    async def clear(self) -> None:
+        """
+        Clear all memory.
+        """
+        await self.chat_history_storage.delete()
+        await self.memory_storage.delete()
+
+    def state_dict(self) -> dict:
```

Comment:
For `state_dict`, would it be better to return a Pydantic `BaseModel` instead of a raw `dict`? A model could define the expected structure for `max_token`, `chat_history_storage`, and `memory_storage`, provide automatic validation, and improve readability and maintainability.

Reply:
fix